lib/parser/source/tree_rewriter.rb

```diff
@@ -99,22 +95,18 @@ def initialize(source_buffer,
                      crossing_deletions: :accept,
                      different_replacements: :accept,
                      swallowed_insertions: :accept)
-        @diagnostics = Diagnostic::Engine.new
-        @diagnostics.consumer = -> diag { $stderr.puts diag.render }
-
+        @diagnostics = nil
         @source_buffer = source_buffer
         @in_transaction = false
 
-        @policy = {crossing_deletions: crossing_deletions,
-                   different_replacements: different_replacements,
-                   swallowed_insertions: swallowed_insertions}.freeze
-        check_policy_validity
+        @crossing_deletions = check_policy_value(crossing_deletions)
+        @different_replacements = check_policy_value(different_replacements)
+        @swallowed_insertions = check_policy_value(swallowed_insertions)
 
-        @enforcer = method(:enforce_policy)
         # We need a range that would be jugded as containing all other ranges,
         # including 0...0 and size...size:
         all_encompassing_range = @source_buffer.source_range.adjust(begin_pos: -1, end_pos: +1)
-        @action_root = TreeRewriter::Action.new(all_encompassing_range, @enforcer)
+        @action_root = TreeRewriter::Action.new(all_encompassing_range, self)
```

Comment:
Passing `self` as an argument is usually a sign of a bad composition (just like passing a private instance method 😄 )

Also, the argument on the `TreeRewriter::Action` is not an `enforcer` anymore, it's a `tree_rewriter`, and so it introduces a bi-directional dependency.

Reply:
Very legitimate comment. I didn't document it, but I thought that asking for an object responding to `call`, or an object responding to `enforce_policy` was pretty similar. That's also the idea in not renaming it; it's an enforcer and needs responding to a single call `enforce_policy`... It just so happens that `TreeRewriter` responds to `enforce_policy` ;-)

Do you have a suggestion?

Reply:
I don't really like both original/new implementations from this POV. Bi-directional relationships between objects cause troubles from time to time, so I'd try to avoid it.

Previously only a single private method was shared with `Action`, and so it was a smaller violation of encapsulation (but it still was a violation). I'd personally keep the original version only for that reason.

At the same time your implementation looks better to me in terms of types and ifaces. Both implementation have issues, the question is which one has more downsides and I don't see "the only" answer here. Up to you to decide.